src/java_application.py

```diff
@@ -7,24 +7,44 @@
 
 import abc
 
+import ops
+
 
 class JavaApplicationBase(abc.ABC):
```

Comment:
I realise this wasn't introduced in this PR, consider an alternative to abstract base classes which is closer to conforming to duck typing: https://peps.python.org/pep-0544/

Reply:
So using a `Protocol` ?  
I propose to revisit that in another PR (there are some other structural changes that could/will probably be done)

Reply:
That is reasonable